File: localbroadcastutil/src/main/java/com/bonepeople/android/localbroadcastutil/OnDestroyListener.java
package com.bonepeople.android.localbroadcastutil;

import android.content.BroadcastReceiver;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.OnLifecycleEvent;

class OnDestroyListener implements LifecycleObserver {
    private final BroadcastReceiver receiver;

    OnDestroyListener(@NonNull BroadcastReceiver receiver) {
        this.receiver = receiver;
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    public void unregister(@NonNull LifecycleOwner lifecycleOwner) {
        LocalBroadcastUtil.unregisterReceiver(receiver);
        lifecycleOwner.getLifecycle().removeObserver(this);
        if (LocalBroadcastUtil.debugEnable) {
            LocalBroadcastUtil.logger.log("已自动注销 " + receiver.toString());
        }
    }

    @Override
    public int hashCode() {
        return receiver.hashCode();
    }

    @Override
    public boolean equals(@Nullable Object obj) {
        if (obj instanceof OnDestroyListener) {
            return ((OnDestroyListener) obj).receiver.equals(receiver);
        }
        return false;
    }
}

File: localbroadcastutil/src/main/java/com/bonepeople/android/localbroadcastutil/OnDestroyListener.kt
package com.bonepeople.android.localbroadcastutil

import android.content.BroadcastReceiver
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.OnLifecycleEvent

class OnDestroyListener(private val receiver: BroadcastReceiver) : LifecycleObserver {

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    fun unregister(lifecycleOwner: LifecycleOwner) {
        LocalBroadcastUtil.unregisterReceiver(receiver)
        lifecycleOwner.lifecycle.removeObserver(this)
        if (LocalBroadcastUtil.DEBUG) {
            LocalBroadcastUtil.logger.log("已自动注销 $receiver")
        }
    }

    override fun hashCode(): Int {
        return receiver.hashCode()
    }

    override fun equals(other: Any?): Boolean {
        return if (other is OnDestroyListener) {
            other.receiver == receiver
        } else false
    }
}